src/stan_math_backend/Lower_stmt.ml

```diff
@@ -0,0 +1,328 @@
+      [Stmts.if_block (Var "pstream__") (List.map ~f:lower_arg args)]
+  | NRFunApp (CompilerInternal FnReject, args) ->
+      let err_strm_name = "errmsg_stream__" in
+      let stream_decl =
+        VariableDefn
+          (make_variable_defn ~type_:(Type_literal "std::stringstream")
+             ~name:err_strm_name () ) in
+      let throw =
+        Throw
+          (Exprs.fun_call "std::domain_error" [(Var err_strm_name).@!("str")])
+      in
+      let add_to_string e =
+        Expression
+          (fun_call "stan::math::stan_print"
+             [VarRef err_strm_name; lower_expr e] ) in
+      (stream_decl :: List.map ~f:add_to_string args) @ [throw]
+  | NRFunApp (CompilerInternal (FnCheck {trans; var_name; var}), args) ->
+      Option.value_map (check_to_string trans) ~default:[] ~f:(fun check_name ->
+          let function_arg = Expr.Helpers.variable "function__" in
+          Exprs.fun_call
+            ("stan::math::check_" ^ check_name)
+            ( [ lower_expr function_arg; Exprs.literal_string var_name
+              ; lower_expr var ]
+            @ List.map ~f:lower_expr args )
+          |> wrap_e )
+  | NRFunApp (CompilerInternal (FnWriteParam {unconstrain_opt; var}), _) -> (
+      let out = Var "out__" in
+      match
+        (unconstrain_opt, Option.bind ~f:constraint_to_string unconstrain_opt)
+      with
+      (* When the current block or this transformation doesn't require unconstraining,
+         use vanilla write *)
+      | None, _ | _, None -> out.@?(("write", [lower_expr var])) |> wrap_e
+      (* Otherwise, use stan::io::serializer's write_free functions *)
+      | Some trans, Some unconstrain_string ->
+          let unconstrain_args = transform_args trans in
+          let write_fn = "write_free_" ^ unconstrain_string in
+          out.@?((write_fn, lower_exprs (unconstrain_args @ [var]))) |> wrap_e )
+  | NRFunApp (CompilerInternal f, args) ->
+      let fname, extra_args = trans_math_fn f in
+      Exprs.fun_call fname (lower_exprs (extra_args @ args)) |> wrap_e
+  | NRFunApp (StanLib (fname, _, _), args) ->
+      Exprs.fun_call (stan_namespace_qualify fname) (lower_exprs args) |> wrap_e
+  | NRFunApp (UserDefined (fname, suffix), args) ->
+      lower_user_defined_fun fname suffix args |> wrap_e
+  | Skip -> [Semicolon]
+  | IfElse (cond, ifbranch, elsebranch) ->
+      [ IfElse
+          ( lower_bool_expr cond
+          , Stmts.block (lower_statement ifbranch)
+          , Option.map ~f:(Fn.compose Stmts.block lower_statement) elsebranch )
+      ]
+  | While (cond, body) ->
+      [While (lower_bool_expr cond, Stmts.block (lower_statement body))]
+  | For
+      { body=
+          { pattern=
+              Assignment
+                (_, {pattern= FunApp (CompilerInternal (FnReadParam _), _); _})
+          ; _ } as body
+      ; _ } ->
+      lower_statement body
+      (* Skip For loop part, just emit body due to the way FnReadParam emits *)
```

Comment:
This case is never reached in our tests. (Thank goodness, I don't want to think about that comment.)

Reply:
I was too fearful of that comment to try touching it, but you're correct in that deleting it does not change any output. Phew.